barretenberg/cpp/pil/avm/gadgets/merkle_tree.pil

```diff
@@ -0,0 +1,76 @@
+include "./poseidon2_full.pil";
+
+// Handles membership and insertion into a merkle tree
+namespace merkle_tree(256);
+    
+    pol commit sel_merkle_tree;
+    sel_merkle_tree * (1 - sel_merkle_tree) = 0;
+    // Will probably used in permutation with other traces
+    pol commit sel_update_op;
+    sel_update_op * (1 - sel_update_op) = 0;
+    pol commit sel_membership_op;
+    sel_membership_op * (1 - sel_membership_op) = 0;
+    // Gotta stop using clk for things that are more like foreign keys
+    pol commit clk;
+    // Inputs to the gadget
+    pol commit leaf_value;
+    pol commit leaf_index;
+    pol commit path_len;
+    pol commit expected_tree_root;
+    // Output of the gadget
+    // Boolean result of the membership check
+    pol commit is_member;
+    is_member * (1 - is_member) = 0;
+
+    // These are all hinted
+    pol commit sibling_value;
+
+    // If we are not done, the path_len decrements by 1
+    sel_merkle_tree * (1 - latch) * (path_len' - path_len + 1) = 0;
+
+    pol commit latch;
+    latch * (1 - latch) = 0;
+    pol commit path_len_inv;
+    // latch == 1 when the path_len == 0
+    sel_merkle_tree * (path_len * (latch * (1 - path_len_inv) + path_len_inv) - 1 + latch) = 0;
+
+    pol commit leaf_index_is_even;
+    leaf_index_is_even * (1 - leaf_index_is_even) = 0;
```

Comment:
I'll add the TODO but it think it might be covered by the "halving" constraint if the initial leaf index is externally constrained (i.e. a public input)